README.md
# sails-service-image

![Build Status](https://img.shields.io/travis/ghaiklor/sails-service-image.svg)
![Coverage](https://img.shields.io/coveralls/ghaiklor/sails-service-image.svg)
![Downloads](https://img.shields.io/npm/dm/sails-service-image.svg)
![Downloads](https://img.shields.io/npm/dt/sails-service-image.svg)
![npm version](https://img.shields.io/npm/v/sails-service-image.svg)
![dependencies](https://img.shields.io/david/ghaiklor/sails-service-image.svg)
![dev dependencies](https://img.shields.io/david/dev/ghaiklor/sails-service-image.svg)
![License](https://img.shields.io/npm/l/sails-service-image.svg)

Service for Sails framework with image features.

## List of supported image processors

- GraphicsMagick
- ImageMagick

## Getting Started

Install this module.

```shell
npm install sails-service-image
```

Then require it in your service and create image instance.

```javascript
// api/services/ImageService.js
import ImageService from 'sails-service-image';

export default ImageService('gm');

// api/controllers/ImageController.js
export default {
  crop: function(req, res) {
    ImageService
      .crop(req.param('file'), {
        x: req.param('x'),
        y: req.param('y'),
        width: req.param('width'),
        height: req.param('height')
      })
      .then(StorageService.upload)
      .then(res.ok)
      .catch(res.negotiate);
  }
};
```

## API

Each of image instances has following methods:

### iptc(image)

Get IPTC information from image. Returns Promise.

`image` - {String|Buffer|Stream} From what image you want to get IPTC information.

### resize(image, [config])

Resize image and get buffer. Returns Promise.

`image` - {String|Buffer|Stream} What image you want to resize.

`config` - {Object}

  - `config.width` - {Number|String} Resulting width of image
  - `config.height` - {Number|String} Resulting height of image
  - `config.direction` - {String} In what direction make the resize (see GM docs)
  - `config.format` - {String} Resulting image format

### crop(image, [config])

Crop image and get buffer. Returns Promise.

`image` - {String|Buffer|Stream} What image you want to crop.

`config` - {Object}

  - `config.width` - {Number|String} Resulting width of image
  - `config.height` - {Number|String} Resulting height of image
  - `config.x` - {Number} Start X coordinate for cropping
  - `config.y` - {Number} Start Y coordinate for cropping
  - `config.format` - {String} Resulting image format

### thumbnail(image)

Just shortcut for `crop` and `resize` with predefined options. Creates thumbnail for image and returns Promise.

`image` - {String|Buffer|Stream} For what image you want to create thumbnail.

## Examples

### GraphicsMagick

```javascript
let gm = ImageService('gm');

// Get IPTC information from image
gm
  .iptc('my-image.jpg')
  .then(console.log.bind(console))
  .catch(console.error.bind(console));

// Resize image
gm
  .resize('my-image.jpg', {
    width: '50%'
  })
  .then(StorageService.upload)
  .then(console.log.bind(console))
  .catch(console.error.bind(console))

// Crop image
gm
  .crop('my-image.jpg', {
    x: 100,
    y: 50,
    width: 200
  })
  .then(StorageService.upload)
  .then(console.log.bind(console))
  .catch(console.error.bind(console))

// Create thumbnail
gm
  .thumbnail('my-image.jpg')
  .then(StorageService.upload)
  .then(console.log.bind(console))
  .catch(console.error.bind(console))
```

### ImageMagick

```javascript
let im = ImageService.create('im');

// Get IPTC information from image
im
  .iptc('my-image.jpg')
  .then(console.log.bind(console))
  .catch(console.error.bind(console));

// Resize image
im
  .resize('my-image.jpg', {
    width: '50%'
  })
  .then(StorageService.upload)
  .then(console.log.bind(console))
  .catch(console.error.bind(console))

// Crop image
im
  .crop('my-image.jpg', {
    x: 100,
    y: 50,
    width: 200
  })
  .then(StorageService.upload)
  .then(console.log.bind(console))
  .catch(console.error.bind(console))

// Create thumbnail
im
  .thumbnail('my-image.jpg')
  .then(StorageService.upload)
  .then(console.log.bind(console))
  .catch(console.error.bind(console))
```

## License

The MIT License (MIT)

Copyright (c) 2015 Eugene Obrezkov

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Changelog

## 3.0.1

- General improvements;

## 3.0.0

- Migration from ES5 to ES6 syntax;
- Remove `.create()` method from ImageService. ImageService is a function now;
- Update and pin dependencies;

## 2.0.0

- Initial release;
- Implement support for GraphicsMagick and ImageMagick;
- Implement getting IPTC tags from image;
- Implement resizing image;
- Implement cropping image;


test/unit/index.test.js
import { assert } from 'chai';
import ImageService from '../../src/index';
import GraphicsMagick from '../../src/GraphicsMagick';
import ImageMagick from '../../src/ImageMagick';

describe('ImageService', () => {
  it('Should properly export', () => {
    assert.isFunction(ImageService);
  });

  it('Should properly create image instances', () => {
    assert.instanceOf(ImageService('gm'), GraphicsMagick);
    assert.instanceOf(ImageService('im'), ImageMagick);
  });

  it('Should properly throw error on unrecognized type', () => {
    assert.throws(() => ImageService('NOT_EXISTS'), Error);
  });
});


test/unit/BaseImage.test.js
import { assert } from 'chai';
import BaseImage from '../../src/BaseImage';

describe('BaseImage', () => {
  it('Should properly export', () => {
    assert.isFunction(BaseImage);
  });

  it('Should properly make objects configurable', () => {
    let image = new BaseImage();

    assert.notOk(image.get('foo'));
    assert.instanceOf(image.set('foo', 'bar'), BaseImage);
    assert.instanceOf(image.set('obj', {foo: 'bar'}), BaseImage);
    assert.deepEqual(image.get(), {foo: 'bar', obj: {foo: 'bar'}});
    assert.deepEqual(image.get('obj'), {foo: 'bar'});
    assert.equal(image.get('obj.foo'), 'bar');
    assert.equal(image.get('foo'), 'bar');
  });

  it('Should properly create image with pre-defined config', () => {
    let image = new BaseImage({
      foo: 'bar',
      obj: {
        foo: 'bar'
      }
    });

    assert.equal(image.get('foo'), 'bar');
    assert.equal(image.get('obj.foo'), 'bar');
    assert.deepEqual(image.get('obj'), {foo: 'bar'});
    assert.notOk(image.get('NOT_EXISTS'));
  });

  it('Should properly get/set provider', () => {
    let image = new BaseImage();

    assert.deepEqual(image.getProvider(), {});
    assert.instanceOf(image.setProvider('PROVIDER'), BaseImage);
    assert.equal(image.getProvider(), 'PROVIDER');
  });
});


test/unit/ImageMagick.test.js
import { assert } from 'chai';
import path from 'path';
import sinon from 'sinon';
import ImageMagick from '../../src/ImageMagick';

describe('ImageMagick', () => {
  it('Should properly export', () => {
    assert.isFunction(ImageMagick);
  });

  it('Should properly get IPTC information', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'identify');

    gm
      .iptc(path.resolve(__dirname, '../fixtures/thailand.jpg'))
      .then(iptc => {
        assert.equal(gm.getProvider().prototype.identify.callCount, 5);
        assert.equal(gm.getProvider().prototype.identify.getCall(0).args[0], '%[IPTC:2:05]');
        assert.equal(gm.getProvider().prototype.identify.getCall(1).args[0], '%[IPTC:2:120]');
        assert.equal(gm.getProvider().prototype.identify.getCall(2).args[0], '%[IPTC:2:25]');
        assert.equal(gm.getProvider().prototype.identify.getCall(3).args[0], '%[IPTC:2:15]');
        assert.isFunction(gm.getProvider().prototype.identify.getCall(4).args[0]);
        assert.deepEqual(iptc, {
          title: '',
          description: '',
          keywords: '',
          category: '',
          format: 'JPEG'
        });

        gm.getProvider().prototype.identify.restore();

        done();
      })
      .catch(done);
  });

  it('Should properly reject on getting IPTC information', done => {
    let gm = new ImageMagick();

    sinon.stub(gm.getProvider().prototype, 'identify', (section, cb) => section instanceof Function ? section('ERROR') : cb('ERROR'));

    gm
      .iptc(path.resolve(__dirname, '../fixtures/thailand.jpg'))
      .then(done)
      .catch(error => {
        assert.equal(error, 'ERROR');
        assert.equal(gm.getProvider().prototype.identify.callCount, 5);
        assert.equal(gm.getProvider().prototype.identify.getCall(0).args[0], '%[IPTC:2:05]');
        assert.equal(gm.getProvider().prototype.identify.getCall(1).args[0], '%[IPTC:2:120]');
        assert.equal(gm.getProvider().prototype.identify.getCall(2).args[0], '%[IPTC:2:25]');
        assert.equal(gm.getProvider().prototype.identify.getCall(3).args[0], '%[IPTC:2:15]');
        assert.isFunction(gm.getProvider().prototype.identify.getCall(4).args[0]);

        gm.getProvider().prototype.identify.restore();

        done();
      });
  });

  it('Should properly resize image with default options', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'resize');

    gm
      .resize(path.resolve(__dirname, '../fixtures/thailand.jpg'))
      .then(result => {
        assert.instanceOf(result, Buffer);
        assert.ok(gm.getProvider().prototype.resize.calledOnce);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[0], 200);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[1], undefined);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[2], '>');

        gm.getProvider().prototype.resize.restore();

        done();
      })
      .catch(done);
  });

  it('Should properly resize image with custom options', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'resize');

    gm
      .resize(path.resolve(__dirname, '../fixtures/thailand.jpg'), {
        width: 100,
        height: 100,
        direction: '<',
        format: 'png'
      })
      .then(result => {
        assert.instanceOf(result, Buffer);
        assert.ok(gm.getProvider().prototype.resize.calledOnce);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[0], 100);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[1], 100);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[2], '<');

        gm.getProvider().prototype.resize.restore();

        done();
      })
      .catch(done);
  });

  it('Should properly reject on resize image', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'resize');
    sinon.stub(gm.getProvider().prototype, 'toBuffer', cb => cb('ERROR'));

    gm
      .resize(path.resolve(__dirname, '../fixtures/thailand.jpg'))
      .then(done)
      .catch(error => {
        assert.equal(error, 'ERROR');
        assert.ok(gm.getProvider().prototype.resize.calledOnce);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[0], 200);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[1], undefined);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[2], '>');

        gm.getProvider().prototype.resize.restore();
        gm.getProvider().prototype.toBuffer.restore();

        done();
      });
  });

  it('Should properly crop image with default options', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'crop');

    gm
      .crop(path.resolve(__dirname, '../fixtures/thailand.jpg'))
      .then(result => {
        assert.instanceOf(result, Buffer);
        assert.ok(gm.getProvider().prototype.crop.calledOnce);
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[0], '100%');
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[1], '100%');
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[2], 0);
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[3], 0);

        gm.getProvider().prototype.crop.restore();

        done();
      })
      .catch(done);
  });

  it('Should properly crop image with custom options', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'crop');

    gm
      .crop(path.resolve(__dirname, '../fixtures/thailand.jpg'), {
        width: 100,
        height: 100,
        x: 200,
        y: 200,
        format: 'png'
      })
      .then(result => {
        assert.instanceOf(result, Buffer);
        assert.ok(gm.getProvider().prototype.crop.calledOnce);
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[0], 100);
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[1], 100);
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[2], 200);
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[3], 200);

        gm.getProvider().prototype.crop.restore();

        done();
      })
      .catch(done);
  });

  it('Should properly reject on crop image', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'crop');
    sinon.stub(gm.getProvider().prototype, 'toBuffer', cb => cb('ERROR'));

    gm
      .crop(path.resolve(__dirname, '../fixtures/thailand.jpg'))
      .then(done)
      .catch(error => {
        assert.equal(error, 'ERROR');
        assert.ok(gm.getProvider().prototype.crop.calledOnce);
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[0], '100%');
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[1], '100%');
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[2], 0);
        assert.equal(gm.getProvider().prototype.crop.getCall(0).args[3], 0);

        gm.getProvider().prototype.crop.restore();
        gm.getProvider().prototype.toBuffer.restore();

        done();
      });
  });

  it('Should properly create thumbnail', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'resize');

    gm
      .thumbnail(path.resolve(__dirname, '../fixtures/thailand.jpg'))
      .then(result => {
        assert.instanceOf(result, Buffer);
        assert.ok(gm.getProvider().prototype.resize.calledOnce);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[0], 200);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[1], undefined);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[2], '>');

        gm.getProvider().prototype.resize.restore();

        done();
      })
      .catch(done);
  });

  it('Should properly reject on creating thumbnail', done => {
    let gm = new ImageMagick();

    sinon.spy(gm.getProvider().prototype, 'resize');
    sinon.stub(gm.getProvider().prototype, 'toBuffer', cb => cb('ERROR'));

    gm
      .thumbnail(path.resolve(__dirname, '../fixtures/thailand.jpg'))
      .then(done)
      .catch(error => {
        assert.equal(error, 'ERROR');
        assert.ok(gm.getProvider().prototype.resize.calledOnce);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[0], 200);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[1], undefined);
        assert.equal(gm.getProvider().prototype.resize.getCall(0).args[2], '>');

        gm.getProvider().prototype.resize.restore();
        gm.getProvider().prototype.toBuffer.restore();

        done();
      });
  });
});


src/BaseImage.js
import _ from 'lodash';

export default class BaseImage {
  constructor(_config) {
    this._config = {};
    this._provider = {};

    _.assign(this._config, _config);
  }

  /**
   * Get configuration value
   * @param {String} [path]
   * @returns {*}
   */
  get(path) {
    return typeof path === 'undefined' ? this._config : _.get(this._config, path);
  }

  /**
   * Set configuration value
   * @param {String} path
   * @param {*} value
   * @returns {BaseImage}
   */
  set(path, value) {
    _.set(this._config, path, value);
    return this;
  }

  /**
   * Get image provider
   * @returns {*}
   */
  getProvider() {
    return this._provider;
  }

  /**
   * Set new provider
   * @param {*} provider
   * @returns {BaseImage}
   */
  setProvider(provider) {
    this._provider = provider;
    return this;
  }
}


src/GraphicsMagick.js
import gm from 'gm';
import Promise from 'bluebird';
import BaseImage from './BaseImage';

export default class GraphicsMagick extends BaseImage {
  constructor(...args) {
    super(...args);

    this.setProvider(gm);
  }

  /**
   * Get title from IPTC
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getTitle(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify('%[IPTC:2:05]', (error, metadata) => error ? reject(error) : resolve(metadata.trimRight()));
    });
  }

  /**
   * Get description from IPTC
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getDescription(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify('%[IPTC:2:120]', (error, metadata) => error ? reject(error) : resolve(metadata.trimRight()));
    });
  }

  /**
   * Get keywords from IPTC
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getKeywords(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify('%[IPTC:2:25]', (error, metadata) => error ? reject(error) : resolve(metadata.trimRight().split(';').join(',')));
    });
  }

  /**
   * Get category from IPTC
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getCategory(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify('%[IPTC:2:15]', (error, metadata) => error ? reject(error) : resolve(metadata.trimRight()));
    });
  }

  /**
   * Get image format
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getFormat(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify((error, data) => error ? reject(error) : resolve(data.format));
    });
  }

  /**
   * Get IPTC information from image
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  iptc(image) {
    return Promise.all([
      this.getTitle(image),
      this.getDescription(image),
      this.getKeywords(image),
      this.getCategory(image),
      this.getFormat(image)
    ]).spread((title, description, keywords, category, format) => Promise.resolve({
      title,
      description,
      keywords,
      category,
      format
    }));
  }

  /**
   * Resize image
   * @param {String|Buffer|Stream} image Source image
   * @param {Object} [_config] Configuration object
   * @returns {Promise}
   */
  resize(image, _config) {
    let config = _config || {};

    return new Promise((resolve, reject) => {
      this.getProvider()(image)
        .resize(config.width || 200, config.height, config.direction || '>')
        .setFormat(config.format || 'jpg')
        .toBuffer((error, buffer) => error ? reject(error) : resolve(buffer));
    });
  }

  /**
   * Crop image
   * @param {String|Buffer|Stream} image Source image
   * @param {Object} [_config] Configuration object
   * @returns {Promise}
   */
  crop(image, _config) {
    let config = _config || {};

    return new Promise((resolve, reject) => {
      this.getProvider()(image)
        .crop(config.width || '100%', config.height || '100%', config.x || 0, config.y || 0)
        .setFormat(config.format || 'jpg')
        .toBuffer((error, buffer) => error ? reject(error) : resolve(buffer));
    });
  }

  /**
   * Create thumbnail for image
   * @param {String|Buffer|Stream} image Source image
   * @param {Object} [_config]
   * @returns {Promise}
   */
  thumbnail(image, _config) {
    return this.resize(image, _config);
  }
}


src/ImageMagick.js
import gm from 'gm';
import Promise from 'bluebird';
import BaseImage from './BaseImage';

export default class GraphicsMagick extends BaseImage {
  constructor(...args) {
    super(...args);

    this.setProvider(gm.subClass({imageMagick: true}));
  }

  /**
   * Get title from IPTC
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getTitle(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify('%[IPTC:2:05]', (error, metadata) => error ? reject(error) : resolve(metadata.trimRight()));
    });
  }

  /**
   * Get description from IPTC
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getDescription(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify('%[IPTC:2:120]', (error, metadata) => error ? reject(error) : resolve(metadata.trimRight()));
    });
  }

  /**
   * Get keywords from IPTC
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getKeywords(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify('%[IPTC:2:25]', (error, metadata) => error ? reject(error) : resolve(metadata.trimRight().split(';').join(',')));
    });
  }

  /**
   * Get category from IPTC
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getCategory(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify('%[IPTC:2:15]', (error, metadata) => error ? reject(error) : resolve(metadata.trimRight()));
    });
  }

  /**
   * Get image format
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  getFormat(image) {
    return new Promise((resolve, reject) => {
      this.getProvider()(image).identify((error, data) => error ? reject(error) : resolve(data.format));
    });
  }

  /**
   * Get IPTC information from image
   * @param {String|Buffer|Stream} image Source image
   * @returns {Promise}
   */
  iptc(image) {
    return Promise.all([
      this.getTitle(image),
      this.getDescription(image),
      this.getKeywords(image),
      this.getCategory(image),
      this.getFormat(image)
    ]).spread((title, description, keywords, category, format) => Promise.resolve({
      title,
      description,
      keywords,
      category,
      format
    }));
  }

  /**
   * Resize image
   * @param {String|Buffer|Stream} image Source image
   * @param {Object} [_config] Configuration object
   * @returns {Promise}
   */
  resize(image, _config) {
    let config = _config || {};

    return new Promise((resolve, reject) => {
      this.getProvider()(image)
        .resize(config.width || 200, config.height, config.direction || '>')
        .setFormat(config.format || 'jpg')
        .toBuffer((error, buffer) => error ? reject(error) : resolve(buffer));
    });
  }

  /**
   * Crop image
   * @param {String|Buffer|Stream} image Source image
   * @param {Object} [_config] Configuration object
   * @returns {Promise}
   */
  crop(image, _config) {
    let config = _config || {};

    return new Promise((resolve, reject) => {
      this.getProvider()(image)
        .crop(config.width || '100%', config.height || '100%', config.x || 0, config.y || 0)
        .setFormat(config.format || 'jpg')
        .toBuffer((error, buffer) => error ? reject(error) : resolve(buffer));
    });
  }

  /**
   * Create thumbnail for image
   * @param {String|Buffer|Stream} image Source image
   * @param {Object} [_config]
   * @returns {Promise}
   */
  thumbnail(image, _config) {
    return this.resize(image, _config);
  }
}
